### PR TITLE
Porting MACI's `FreeForAll` and `SignUpToken` gatekeepers

### PR DESCRIPTION
**Pull request by 0xjei**
## Description

This PR introduces `FreeForAll` and `SignUpToken` MACI's gatekeepers as Excubia w/ some new features:
- [x] [FreeForAllGatekeeper](https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/contracts/gatekeepers/FreeForAllSignUpGatekeeper.sol) (added logic to avoid passing twice with the same address).
- [x] [SignUpTokenGatekeeper](https://github.com/privacy-scaling-explorations/maci/blob/dev/contracts/contracts/gatekeepers/SignUpTokenGatekeeper.sol) (evolved to support whatever ERC721 contract).

## Related Issue(s)
see #18 
needs #17 

## Checklist

-   [x] My code follows the style guidelines of this project
-   [x] I have performed a self-review of my code
-   [x] I have commented my code, particularly in hard-to-understand areas
-   [x] My changes generate no new warnings
-   [x] I have run `yarn format` and `yarn compile` without getting any errors
-   [x] I have added tests that prove my fix is effective or that my feature works
-   [x] New and existing unit tests pass locally with my changes

### CODE REVIEW

File: packages/excubiae/contracts/extensions/ERC721Excubia.sol [hunk 1]
```diff
@@ -0,0 +1,60 @@
...
+        registeredTokenIds[tokenId] = true;
+    }
+
+    /// @notice Internal function to handle the gate protection (token ownership check) logic.
+    /// @dev Checks if the passerby is the owner of the token.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Additional data required for the check (e.g., encoded token ID).
+    /// @return True if the passerby owns the token, false otherwise.
+    function _check(address passerby, bytes calldata data) internal view override returns (bool) {
+        super._check(passerby, data);
```
Comment (ctrlc03) [inline packages/excubiae/contracts/extensions/ERC721Excubia.sol:51]: not sure this is needed, it's an empty function in the abstract contract

File: packages/excubiae/contracts/extensions/FreeForAllExcubia.sol [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+        if (registeredPassersby[passerby]) revert AlreadyRegistered();
+
+        registeredPassersby[passerby] = true;
+    }
+
+    /// @notice Internal function to handle the gate protection logic.
+    /// @dev This function always returns true, signaling that any passerby is able to pass the gate.
+    /// @return True, allowing any passerby to pass the gate.
+    function _check(address passerby, bytes calldata data) internal view override returns (bool) {
+        super._check(passerby, data);
```
Comment (ctrlc03) [inline packages/excubiae/contracts/extensions/FreeForAllExcubia.sol:34]: same as above, not sure this is needed?

File: packages/excubiae/contracts/extensions/ERC721Excubia.sol [hunk 3]
```diff
@@ -0,0 +1,60 @@
...
+    /// @dev Calls the parent `_pass` function and registers the NFT ID to avoid passing the gate twice.
+    /// @param passerby The address of the entity attempting to pass the gate.
+    /// @param data Additional data required for the check (e.g., encoded token ID).
+    function _pass(address passerby, bytes calldata data) internal override {
+        super._pass(passerby, data);
+
+        uint256 tokenId = abi.decode(data, (uint256));
+
+        // Avoiding passing the gate twice with the same token ID.
+        if (registeredTokenIds[tokenId]) revert AlreadyRegistered();
```
Comment (ctrlc03) [inline packages/excubiae/contracts/extensions/ERC721Excubia.sol:40]: wonder if we should fail first? i.e. check if it's already registered before calling _check(via super._pass)

**Review by ctrlc03 — COMMENTED** [review]
🚀 let's some comments

File: packages/excubiae/contracts/extensions/ERC721Excubia.sol (hunk 3 above)
Comment (0xjei) [reply packages/excubiae/contracts/extensions/ERC721Excubia.sol:40]: yes, we can save some gas for the user in case of failing - correct

**Review by ctrlc03 — APPROVED** [review]
Thanks!!

**Pull request merged**